src/Discord.Net.Core/Utils/Cacheable.cs

```diff
@@ -0,0 +1,60 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Discord
+{
+    /// <summary>
+    /// Contains an entity that may be cached.
+    /// </summary>
+    /// <typeparam name="TEntity">The type of entity that is cached</typeparam>
+    /// <typeparam name="TId">The type of this entity's ID</typeparam>
+    public struct Cacheable<TEntity, TId>
+        where TEntity : IEntity<TId>
+        where TId : IEquatable<TId>
+    {
+        /// <summary>
+        /// Is this entity cached?
+        /// </summary>
+        public bool HasValue { get; }
```

Comment:
Shouldn't the value types be wrapped in a nullable anyway? Probably making a bigger deal out of this than necessary, but it seems worrying for the potential of HasValue to be false if Value actually exists.